agent/config/file_watcher.go

```diff
@@ -81,16 +81,38 @@ func (w *FileWatcher) Stop() error {
 	w.stopOnce.Do(func() {
 		w.cancel()
 		<-w.done
-		close(w.EventsCh)
 		err = w.watcher.Close()
 	})
 	return err
 }
 
-func (w *FileWatcher) add(filename string) error {
-	if isSymLink(filename) {
-		return fmt.Errorf("symbolic links are not supported %s", filename)
+// Add a file to the file watcher
+// Add will lock the file watcher during the add
+func (w *FileWatcher) Add(filename string) error {
+	return w.add(filename)
+}
+
+// Remove a file from the file watcher
+// Remove will lock the file watcher during the remove
+func (w *FileWatcher) Remove(filename string) {
+	w.configFilesLock.Lock()
+	defer w.configFilesLock.Unlock()
+	delete(w.configFiles, filename)
+}
+
+// Replace a file in the file watcher
+// Replace will lock the file watcher during the replace
+func (w *FileWatcher) Replace(oldFile, newFile string) error {
+	if oldFile == newFile {
+		return nil
 	}
+	w.configFilesLock.Lock()
+	defer w.configFilesLock.Unlock()
+	delete(w.configFiles, oldFile)
```

Comment:
`Replace` calls `add()` at the bottom, which also locks, then this would deadlock itself.

Either make `Replace` non-atomic  (lock, delete, unlock, lock, add, unlock) or refactor Replace and add together so that you only lock once.